cpp-ch/local-engine/Storages/Output/NormalFileWriter.cpp

```diff
@@ -19,13 +19,134 @@
 #include <QueryPipeline/QueryPipeline.h>
 #include <Poco/URI.h>
 #include <Common/DebugUtils.h>
+#include <Columns/ColumnConst.h>
+#include <Columns/ColumnArray.h>
+#include <Columns/ColumnMap.h>
 
 namespace local_engine
 {
 
+using namespace DB;
+
 const std::string SubstraitFileSink::NO_PARTITION_ID{"__NO_PARTITION_ID__"};
 const std::string SparkPartitionedBaseSink::DEFAULT_PARTITION_NAME{"__HIVE_DEFAULT_PARTITION__"};
 
+/// For Nullable(Map(K, V)) or Nullable(Array(T)), if the i-th row is null, we must make sure its nested data is empty.
+/// It is for ORC/Parquet writing compatiability. For more details, refer to
+/// https://github.com/apache/incubator-gluten/issues/8022 and https://github.com/apache/incubator-gluten/issues/8021
+static ColumnPtr truncateNestedArrayOrMapIfNull(const ColumnPtr & column)
```

Comment:
better rename function to `truncateNestedDataIfNull`, since not only array or map, also include tuple

Reply:
Essentially, the truncate to the tuple is actually the truncate to the map/array nested in the tuple. So the original name is more appropriate

Reply:
fixed.